File: puzzles/ruby/mime-type.rb
# For each of the Q filenames, display on a line the corresponding MIME type.
# If there is no corresponding type, then display UNKNOWN.

hash = Hash.new
fileExtension = ""
mimeType = ""
dotIndex = 0

@pairCount = gets.to_i # Number of elements which make up the association table.
@filesCount = gets.to_i # Number Q of file names to be analyzed.

@pairCount.times do
  # ext: file extension
  # mt: MIME type.
  ext, mt = gets.split(" ")
  ext = ext.downcase
  hash.store(ext, mt)
end

#Test
# hash.each do |ext, mt|
#   STDERR.puts "#{ext} #{mt}"
# end

@filesCount.times do
  fileName = gets.chomp # One file name per line.
  #STDERR.puts(fileName)
  dotIndex = fileName.rindex('.')
  #STDERR.puts(dotIndex)
  if dotIndex == nil
    puts "UNKNOWN"
  elsif (dotIndex >= 0) && (dotIndex < (fileName.length - 1))
    fileExtension = fileName[dotIndex + 1 .. -1].downcase
    #STDERR.puts(fileExtension)
    mimeType = hash.fetch(fileExtension, 'UNKNOWN')
    puts(mimeType)
  else
    puts "UNKNOWN"
  end
end

File: puzzles/js/mime-type.js
'use strict';

let map = {};

const pairCount = parseInt(readline()); // Number of elements which make up the association table.
const fileCount = parseInt(readline()); // Number of file names to be analyzed.
for (let i = 0; i < pairCount; ++i) {
  let inputs = readline().split(' ');
  let extension = inputs[0].toLowerCase(); // file extension
  let type = inputs[1]; // MIME type
  map[extension] = type;
}

// Debug
//for (let key in map)
//  printErr(key + '=' + map[key]);

// For each of the filenames, display on a line the corresponding MIME type.
// If there is no corresponding type, then display UNKNOWN.
for (let i = 0; i < fileCount; ++i) {
  let fileName = readline(); // One file name per line.
  let dotIndex = fileName.lastIndexOf('.');
  if ((dotIndex === -1) || (dotIndex === (fileName.length - 1))) {
    print('UNKNOWN');
  } else {
    let fileExt = fileName.substring(dotIndex + 1, fileName.length).toLowerCase();
    let mimeType = map[fileExt];
    if (mimeType === undefined)
      print('UNKNOWN');
    else
      print(mimeType);
  }
}
